### PR TITLE
Issue #195: Compile mousetrap only on Windows

### PR DESCRIPTION
**Pull request by magiconair**
* Create command_win.go and command_notwin.go for windows only code
* Move call to mousetrap hook into separate preExecHook() function

### CODE REVIEW

File: cobra.go [hunk 1]
```diff
@@ -40,14 +40,6 @@ var initializers []func()
 // Set this to true to enable it
 var EnablePrefixMatching bool = false
 
-// enables an information splash screen on Windows if the CLI is started from explorer.exe.
-var EnableWindowsMouseTrap bool = true
```
Comment (eparis) [inline cobra.go]: I have no idea if anyone was using it, but this means you can't turn off the mousetrap/exit, right?

File: command.go [hunk 2]
```diff
@@ -626,13 +623,8 @@ func (c *Command) ExecuteC() (cmd *Command, err error) {
...
 
-	if EnableWindowsMouseTrap && runtime.GOOS == "windows" {
-		if mousetrap.StartedByExplorer() {
-			c.Print(MousetrapHelpText)
-			time.Sleep(5 * time.Second)
-			os.Exit(1)
-		}
-	}
+	// windows hook
+	preExecHook(c)
```
Comment (eparis) [inline command.go]: It's bikeshedding, but we already have user exposed things called PreExec stuff. I'm trying to think of a better name, but I don't love this one.

File: cobra.go (hunk 1 above)
Comment (magiconair) [inline cobra.go]: You can replace the code with an empty function. Now that I think about it it should be more like this (sans better name. Don't like it either):

```
var preExecHook = preExecHookWin

func preExecHookWin() {...}
```

This way a user can override the function with a custom one - or replace it with nil to disable it.

File: command.go (hunk 2 above)
Comment (magiconair) [inline command.go]: yeah, don't like it either. Feel free to be more creative :)

**Comment by magiconair** [comment]
@eparis pushed a version which has uses the function var approach. Did a force push on the previous version (used to gerrit). Not sure whether I messed something up.

File: cobra.go (hunk 1 above)
Comment (eparis) [inline cobra.go]: 1. preExecHook is a private function, so I can't overwrite it from outside
2. This is an API change vs the variable. So while I think this is a better API, I'd rather not redo it unless we need to.

**Comment by spf13** [comment]
@eparis . I think this latest push looks good and the approach is sound. Do you want to review also since you provided feedback earlier?

**Comment by eparis** [comment]
I like the approach, but people can't turn it off. (Which is fine if @spf13 is fine with it)

I'm not allowed to call   `cobra.preExecHook = myFunc` but I was allowed to do `cobra.EnableWindowsMouseTrap = false`

**Comment by magiconair** [comment]
The reason I made the preExecHook private was that I felt that this isn't something people should hook into. So it is more an implementation detail than a public entry point. Hence also the somewhat generic name. I agree that disabling the functionality on Windows is no longer possible but that I could fix by reintroducing the flag.

**Comment by spf13** [comment]
I can't think of why someone would turn it off. I suppose we can ship
without the flag and add it if someone asks for it? Less configuration is a
good thing.

**Pull request merged**